### PR TITLE
Fix usage of namespace shared informers in existing admission controllers

### PR DESCRIPTION
**Pull request by derekwaynecarr**
Fixes https://github.com/kubernetes/kubernetes/issues/29473

/cc @liggitt @ncdc

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
This is WIP (it doesnt work), opening it so I can reference this PR in the interim PR that disables the test to stop bleeding.

**Comment by derekwaynecarr** [comment]
@deads2k - this fixes the flakes, and ensures the existing handlers that used shared informer are waiting for sync.  Note that I moved some stuff into admission.Handler to consolidate that logic by letting a user latently register a `ReadyFunc`.  I think this works with the pattern you have used in downstream plug-ins.  PTAL.

@hodovska - the key thing to move in a future PR is the `NamespaceLifecycle` handler.

File: pkg/admission/handler.go [hunk 1]
```diff
@@ -42,3 +54,32 @@ func NewHandler(ops ...Operation) *Handler {
...
+	}
+	return h.waitForReadyInternal(time.After(timeToWaitForReady))
+}
+
+func (h *Handler) waitForReadyInternal(timeout <-chan time.Time) bool {
+	// there is no configured ready func, so return immediately
+	if h.readyFunc == nil {
+		return true
+	}
+	for !h.readyFunc() {
```
Comment (ncdc) [inline pkg/admission/handler.go]: Instead of polling, what if `readyFunc` returned a `chan struct{}` that you could select on?

Comment (derekwaynecarr) [inline pkg/admission/handler.go]: this is largely done this way to be consistent with origin.  i am not sure how fancy we need to be.

File: pkg/admission/handler.go [hunk 2]
```diff
@@ -42,3 +54,32 @@ func NewHandler(ops ...Operation) *Handler {
 		operations: operations,
 	}
 }
+
+// SetReadyFunc allows late registration of a ReadyFunc to know if the handler is ready to process requests.
+func (h *Handler) SetReadyFunc(readyFunc ReadyFunc) {
```
Comment (deads2k) [inline pkg/admission/handler.go]: We have places that rely on multiple caches.  You want them unioned separately or would you be ok make this `AddReadyFunc` and waiting for all of them?

File: pkg/admission/handler.go (hunk 1 above)
Comment (deads2k) [inline pkg/admission/handler.go]: > this is largely done this way to be consistent with origin. i am not sure how fancy we need to be.

I'm fine with the naive impl for now.

**Comment by deads2k** [comment]
Optional nit.  lgtm otherwise.

**Comment by deads2k** [comment]
marked priority to match bug priority.

**Comment by deads2k** [comment]
@k8s-bot test this: issue #27462

File: pkg/admission/handler.go (hunk 2 above)
Comment (derekwaynecarr) [inline pkg/admission/handler.go]: i am fine either way.

**Comment by derekwaynecarr** [comment]
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/27462

**Comment by deads2k** [comment]
@k8s-bot test this issue #27462

**Comment by derekwaynecarr** [comment]
@k8s-bot test this issue #27462

**Pull request merged**